File: aoc/2015/day-4/the-ideal-stocking-stuffer.js
'use strict';
// 2024-12-27 10:48:13
// https://adventofcode.com/2015/day/4
import { isMainThread, Worker, workerData } from 'node:worker_threads';
import { getFilePath, md5, summonTheFile } from '../../util/util.js';

const inputFile = 'sample.txt';
const cufileName = 'the-ideal-stocking-stuffer.js';
const filePath = getFilePath(import.meta.url, inputFile);
const fileContents = await summonTheFile(filePath);
const zeroLen = 6;
const zeros = '000000';
const max = 10000000;

function matchHash(start, end) {
  for (let index = start; index < end; index++) {
    const hash = md5(fileContents + index);
    if (hash.substring(0, zeroLen) === zeros) {
      console.log(hash, index);
      return index;
    }
  }
}

const threads = new Set();
function splitToParts(num, threadCount = 1) {
  const parts = [];
  const chunkSize = Math.ceil(num / threadCount);

  for (let index = 0; index < num; index += chunkSize) {
    const end = Math.min(index + chunkSize, num);
    parts.push({ start: index, end });
  }
  return parts;
}

if (isMainThread) {
  const result = [];
  const parts = splitToParts(max, 10).toSorted((a, b) => a - b);
  for (const p of parts) {
    threads.add(
      new Worker(getFilePath(import.meta.url, cufileName), {
        workerData: p,
      })
    );
  }
  threads.forEach((thread) => {
    thread.on('error', (err) => {
      throw err;
    });
    thread.on('exit', () => {
      threads.delete(thread);
      console.log(`Thread exiting, ${threads.size} running...`);
    });
    thread.on('message', (msg) => {
      result.push(msg);
    });
  });
  console.log(result);
} else {
  matchHash(workerData.start, workerData.end);
}
// main();
// p1 - 346386
// p2 - 9958218

File: aoc/2015/day-4/the-ideal-stocking-stuffer.go
package day4_2015

// 2024-12-27 12:34:08
// https://adventofcode.com/2015/day/4
import (
	"advent/code/util"
	"log"
	"strconv"
	"strings"
)

var zeros = "00000"
var zeros2 = "000000"
var max = 10000000

func Main() int {

	val, err := util.GetFileContents("/2015/day-4/sample.txt")
	if err != nil {
		log.Fatal(err)
	}
	fileContents := strings.TrimSpace(string(val[:]))
	return matchHash(0, max, fileContents, zeros2)

}

// p1 - 609043
// p2 - 9958218
func matchHash(start, end int, str, prefix string) int {

	for i := start; i < end; i++ {

		hash := util.Md5(str + strconv.Itoa(i))

		if strings.HasPrefix(string(hash), prefix) {
			return i
		}
	}
	return 0
}
